woocommerce/woocommerce-ios
# Improve application log viewer

## Why

The old `ApplicationLogDetailViewController` has not kept up with the times. The current implementation is not that different from [the initial one in 2013](https://github.com/wordpress-mobile/WordPress-iOS/blob/037f7469e07985c2c874299c06a9d043f76e958e/WordPress/Classes/ActivityLogDetailViewController.m), and it has some issues:
- Most of the times that you go to that view, you either want to share the log, or read the latest lines
- The scroll view doesn't extend to the full width, so it's hard to drag the scroll bar
- Some layout guides might be wrong because dragging the scroll bar (when you can grab it), doesn't always get you to the bottom. I've noticed this more often in the WordPress app, where logs tend to be much longer.
- When you load the view, the text view is not scrolled at the top, it has a small offset.


https://user-images.githubusercontent.com/8739/121161371-36103180-c84d-11eb-8937-85006f34e512.mov

## What changed

Instead of dumping all the text into a text view, the new log viewer uses SwiftUI's List to show each line. We do a little bit of parsing to extract the log timestamp into a smaller header, so the log message is front and center. I think this will also make it easier to implement a searching feature if that seems useful enough.

New list design | Scroll to bottom | Button hides when not needed
-|-|-
<img src="https://user-images.githubusercontent.com/8739/121161518-5e982b80-c84d-11eb-80ff-199311b943b4.png" width="564">|![scroll-to-bottom](https://user-images.githubusercontent.com/8739/121162627-63a9aa80-c84e-11eb-9828-68fe1f27a74c.gif)|![button-appear](https://user-images.githubusercontent.com/8739/121162636-64dad780-c84e-11eb-94b9-3166edd73752.gif)

## How

Implementing this with SwiftUI was relatively easy, but reading the scroll position is still a little rough. I had to use the `onAppear`/`onDisappear` modifiers on cells to keep track of whether the last cell was visible. This method is not completely reliable as they often get fired more than necessary (it seems there's a fix on iOS 15 for this, see paNNhX-et-p2), and the `onDisappear` method seems to be called when the cell is disposed/reused, not necessarily the moment it goes off-screen. The first problem was solved by using Combine and debouncing events each 100ms. The second problem isn't solved, but the result for this particular use case isn't too bad: the button doesn't hide exactly when the last cell scrolls off-screen, but it will hide eventually as you keep scrolling up.

## To test

1. Go to App Settings > Help & Support > View Application Log.
2. Pick the current log or any other.
3. The log should load with the new view scrolled to the top
4. Tapping on the arrow down button should scroll to the end of the log
5. Dragging the scroll bar should work as expected
6. Tapping the share button should allow to share the log (I tested with Airdrop)

Update release notes:
